Firebase/Messaging/FIRMessaging.m

```diff
@@ -955,18 +956,14 @@ - (void)didReceiveDefaultInstanceIDToken:(NSNotification *)notification {
 - (void)defaultInstanceIDTokenWasRefreshed:(NSNotification *)notification {
   // Retrieve the Instance ID default token, and if it is non-nil, post it
   NSString *token = self.instanceID.token;
-  // Sometimes Instance ID doesn't yet have a token, so wait until the default
-  // token is fetched, and then notify. This ensures that this token should not
-  // be nil when the developer accesses it.
-  if (token != nil) {
-    NSString *oldToken = self.defaultFcmToken;
-    self.defaultFcmToken = [token copy];
-    if (self.defaultFcmToken && ![self.defaultFcmToken isEqualToString:oldToken]) {
-      [self notifyDelegateOfFCMTokenAvailability];
-    }
-    NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
-    [center postNotificationName:FIRMessagingRegistrationTokenRefreshedNotification object:nil];
+  NSString *oldToken = self.defaultFcmToken;
+  self.defaultFcmToken = [token copy];
+  if ((self.defaultFcmToken && ![self.defaultFcmToken isEqualToString:oldToken]) ||
+      (self.defaultFcmToken && !oldToken) || (!self.defaultFcmToken && oldToken)) {
+    [self notifyDelegateOfFCMTokenAvailability];
   }
+  NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
+  [center postNotificationName:FIRMessagingRegistrationTokenRefreshedNotification object:nil];
```

Comment:
Is it intentional to always post here? There are no cases left where there shouldn't be a notification?

Reply:
Good catch! it should be called any time the delgate method is called, moving into the notify delegate method. 
I didn't test the case for notification only for delegate method. I will update with a integration test.

Reply:
Looks like we need to encrypt a googleService-Info.pilst file to ensure the network call succeed in integration test. I might have to add it in another PR as I haven't figure out how to do that.